File: asset/js/charts/laicite/actors.js
/**
 * IWAC Visualizations — Laïcité block: Actors & institutions (issue #14, view 7).
 *
 * Who is speaking laïcité, in which decade, and in which country. Built from
 * the curated authority records co-occurring with the dossier's items, so
 * every row is a real catalogued entity with its own page rather than a
 * string harvested from text.
 *
 * Two panels:
 *   - a decade heatmap over the top actors, which is what answers "who was
 *     speaking laïcité *then*" — the ranked list alone flattens sixty years
 *     into one column and hides every succession;
 *   - the ranked list itself, where each row links to the entity's record
 *     and carries its span, its corpora and its tagged share.
 *
 * Counted once per item throughout: an organisation named three times in
 * one article is one article's worth of evidence.
 */
(function () {
    'use strict';

    var ns = window.IWACVis;
    if (!ns || !ns.panels) {
        console.warn('IWACVis.laicite actors: missing panels — check load order');
        return;
    }
    var P = ns.panels;
    var L = ns.laicite = ns.laicite || {};

    /** Authority types, in the order the filter offers them. */
    L.ACTOR_TYPES = ['Personnes', 'Organisations', 'Événements'];

    /** How many actors the heatmap can carry before it stops being readable. */
    var HEATMAP_ROWS = 20;

    /** The type keys actually present in the bundle, in canonical order. */
    L.actorTypes = function (bundle) {
        var present = {};
        ((bundle || {}).actors || []).forEach(function (a) {
            present[a.type] = true;
        });
        return L.ACTOR_TYPES.filter(function (t) { return present[t]; });
    };

    function filtered(bundle, state) {
        var actors = (bundle || {}).actors || [];
        if (!state.actorType) return actors;
        return actors.filter(function (a) { return a.type === state.actorType; });
    }

    /**
     * @param {Object} cfg {bundle, state, siteBase}
     * @returns {{root: HTMLElement, mount: function():void,
     *            update: function(Object):void}}
     *
     * `update(state)` is what keeps the type filter from rebuilding this
     * view (Tier 8 / S17). The panel's chrome — heading, description, the
     * method note — and the heatmap's HOST element are built once and stay;
     * only the option and the ranked list change. That is the whole point:
     * a retained host means `registerChart`'s instance is still alive, so
     * the heatmap ANIMATES between filters instead of being disposed and
     * replaced, the theme observer keeps tracking one instance rather than
     * a new one per keystroke, and the scroll position survives because the
     * view host is never emptied.
     */
    L.buildActors = function (cfg) {
        var bundle = cfg.bundle;
        var root = P.el('div', 'iwac-vis-laicite-actors');

        var panel = P.el('div', 'iwac-vis-panel');
        panel.appendChild(P.el('h4', null, P.t('laicite.actors_title')));
        panel.appendChild(P.el('p', 'iwac-vis-panel-desc',
            P.t('laicite.actors_desc')));

        if (!bundle) {
            panel.appendChild(P.buildNoDataState());
            root.appendChild(panel);
            return { root: root, mount: function () {}, update: function () {} };
        }

        // Both slots exist from the start, empty if this filter has nothing
        // to put in them. Creating them on demand would mean a new chart
        // host, which is exactly what `update` exists to avoid.
        var chartEl = P.el('div', 'iwac-vis-chart iwac-vis-laicite-actors-chart');
        var listHost = P.el('div', 'iwac-vis-laicite-actors-list-host');
        panel.appendChild(chartEl);
        panel.appendChild(listHost);

        // The join is a join: some subject strings match no authority record
        // at all, and a panel built on a lookup should say how lossy the
        // lookup was rather than presenting its output as the whole truth.
        var method = P.el('div', 'iwac-vis-laicite-method');
        method.appendChild(P.el('p', null, P.t('laicite.actors_note')));
        method.appendChild(P.el('p', 'iwac-vis-laicite-method-stats',
            P.t('laicite.actors_method', {
                min: bundle.min_items,
                records: P.formatNumber(bundle.index_records || 0),
                unresolved: P.formatNumber(bundle.unresolved_total || 0)
            })));
        panel.appendChild(method);
        root.appendChild(panel);

        /** Fill both slots for one state. Returns the heatmap option, or null. */
        function paint(state) {
            var actors = filtered(bundle, state || {});
            listHost.innerHTML = '';
            if (!actors.length) {
                chartEl.hidden = true;
                listHost.appendChild(P.buildEmptyState('laicite.actors_empty'));
                return null;
            }
            listHost.appendChild(buildActorList(actors, cfg));
            var option = heatmapOption(bundle, actors);
            chartEl.hidden = !option;
            return option;
        }

        var pending = paint(cfg.state);

        return {
            root: root,
            mount: function () {
                if (!pending) return;
                ns.registerChart(chartEl, function (el, instance) {
                    instance.setOption(pending, { notMerge: true });
                });
            },
            update: function (state) {
                var option = paint(state);
                var live = ns.getLiveChart && ns.getLiveChart(chartEl);
                if (!option) return;
                if (live) {
                    // `notMerge: false` so the heatmap keeps its axes and
                    // animates the cells rather than snapping.
                    live.setOption(option, { notMerge: false, lazyUpdate: true });
                    return;
                }
                // No live instance: the first paint had nothing to draw, so
                // nothing was registered. Register now.
                pending = option;
                ns.registerChart(chartEl, function (el, instance) {
                    instance.setOption(pending, { notMerge: true });
                });
            }
        };
    };

    /**
     * Actors × decades, as shares of each actor's own total rather than raw
     * counts. Row-normalised because the question is *when* an actor speaks,
     * not how large it is — the ranked list below already answers size, and
     * on raw counts the three or four biggest records would be the only
     * cells with any colour at all.
     *
     * Returns the OPTION, not a host: the host is owned by `buildActors` so
     * it can survive a filter change (S17).
     */
    function heatmapOption(bundle, actors) {
        var decades = bundle.decades || [];
        if (decades.length < 2) return null;
        var rows = actors.slice(0, HEATMAP_ROWS);
        if (rows.length < 2) return null;
        var C = ns.chartOptions;
        if (!C || !C.heatmapMatrix) return P.emptyChartOption();

        var yLabels = rows.map(function (a) { return a.name; });
        var cells = [];
        rows.forEach(function (actor, y) {
            var total = (actor.by_decade || []).reduce(function (s, n) {
                return s + n;
            }, 0) || 1;
            decades.forEach(function (decade, x) {
                var n = (actor.by_decade || [])[x] || 0;
                cells.push([x, y, Math.round((n / total) * 100), n]);
            });
        });

        return C.heatmapMatrix(
            { xLabels: decades, yLabels: yLabels, cells: cells },
            {
                visualMax: 100,
                cellLabels: false,
                cellBorder: true,
                tooltipFormatter: function (p) {
                    return P.escapeHtml(yLabels[p.value[1]]) + '<br>'
                        + P.escapeHtml(decades[p.value[0]]) + ': <strong>'
                        + P.formatNumber(p.value[3]) + '</strong> ('
                        + p.value[2] + '%)';
                }
            }
        );
    }

    function buildActorList(actors, cfg) {
        var siteBase = cfg.siteBase || '';
        var max = actors.reduce(function (m, a) {
            return Math.max(m, a.items || 0);
        }, 0) || 1;

        var list = P.el('ol', 'iwac-vis-laicite-actor-list');
        actors.forEach(function (actor) {
            var li = P.el('li', 'iwac-vis-laicite-actor');

            var head = P.el('div', 'iwac-vis-laicite-actor-head');
            var name;
            if (siteBase && actor.o_id) {
                name = P.el('a', 'iwac-vis-laicite-actor-name', actor.name);
                name.href = siteBase + '/item/' + actor.o_id;
            } else {
                name = P.el('span', 'iwac-vis-laicite-actor-name', actor.name);
            }
            head.appendChild(name);
            head.appendChild(L.chip(P.t('laicite.actor_type_' + actor.type),
                'is-type'));
            li.appendChild(head);

            var bar = P.el('div', 'iwac-vis-laicite-actor-bar');
            var fill = P.el('span', 'iwac-vis-laicite-actor-fill');
            fill.style.width = Math.max(2, ((actor.items || 0) / max) * 100) + '%';
            bar.appendChild(fill);
            li.appendChild(bar);

            var span = (actor.first_year && actor.last_year)
                ? actor.first_year + '–' + actor.last_year : '—';
            li.appendChild(P.el('p', 'iwac-vis-laicite-actor-meta',
                P.t('laicite.actor_stats', {
                    items: P.formatNumber(actor.items || 0),
                    tagged: P.formatNumber(actor.tagged || 0),
                    span: span
                })));

            // Which corpora carry this actor. A name that only ever appears
            // in the periodicals is a different finding from one the
            // mainstream press also names, and the list is where that shows.
            var corpora = P.el('p', 'iwac-vis-laicite-actor-corpora');
            L.SUBSETS.forEach(function (subset) {
                var n = (actor.by_subset || {})[subset];
                if (!n) return;
                corpora.appendChild(L.chip(
                    L.subsetLabel(subset) + ' ' + P.formatNumber(n), 'is-corpus'));
            });
            if (corpora.childNodes.length) li.appendChild(corpora);

            list.appendChild(li);
        });
        return list;
    }
})();

File: scripts/laicite/actors.py
"""Co-occurring persons and organisations — ``laicite-actors.json``.

Joined against the ``index`` authority records, so an actor is a curated
entity with an o:id rather than a string that happened to be capitalised.
"""
from __future__ import annotations

from collections import Counter
from typing import Any, Dict, List, Optional, Tuple

from iwac_utils import (
    generate_timestamp,
    normalize_location_name,
    parse_coordinates,
    parse_pipe_separated,
)



class ActorsMixin:
    """``LaiciteGenerator``'s actors half. Mixed in by ``laicite.generator``."""


    @staticmethod
    def _decade(year: Optional[int]) -> Optional[str]:
        return f"{year // 10 * 10}s" if year else None

    def _authority_index(self) -> Tuple[Dict[str, Dict[str, Any]], int]:
        """``normalized name → authority record`` over EVERY index type.

        Every type is indexed, not only the ones a caller wants, so that a
        subject string resolving to a ``Sujets`` record counts as resolved
        rather than landing in the unresolved list. Otherwise "unresolved"
        would be dominated by *Laïcité*, *Paix*, *Politique* — the research
        vocabulary, which is catalogued and simply is not an actor.
        """
        df = self._index_records()
        if df is None:
            return {}, 0
        by_name: Dict[str, Dict[str, Any]] = {}
        count = 0
        aliases: List[Tuple[str, Dict[str, Any]]] = []
        for _, row in df.iterrows():
            title = str(row.get("Titre") or "").strip()
            if not title:
                continue
            try:
                o_id = int(row["o:id"])
            except (TypeError, ValueError, KeyError):
                continue
            record = {
                "o_id": o_id,
                "name": title,
                "type": str(row.get("Type") or "").strip(),
                "coords": parse_coordinates(row.get("Coordonnées")),
            }
            count += 1
            by_name[normalize_location_name(title)] = record
            for alt in parse_pipe_separated(row.get("Titre alternatif")):
                key = normalize_location_name(alt)
                if key:
                    aliases.append((key, record))
        # Aliases are applied after every canonical title, so an alternative
        # title can never shadow another record's real one.
        for key, record in aliases:
            by_name.setdefault(key, record)
        return by_name, count

    def build_actors(self) -> Dict[str, Any]:
        """Who is speaking laïcité, and when (issue #14, view 7).

        Joins each dossier item's ``subject`` list against the IWAC index
        and keeps the records that are *actors*: persons, organisations and
        curated events. ``Sujets`` are excluded here — not because they are
        noise (the repo's CLAUDE.md is explicit that they are not) but
        because they answer a different question, and the frame legend and
        the arenas view already answer it.

        Counted once per item: an organisation named three times in one
        article is one item's worth of evidence, not three.
        """
        scans = self.scan_all()
        by_name, indexed = self._authority_index()
        actor_types = {"Personnes", "Organisations", "Événements"}

        stats: Dict[int, Dict[str, Any]] = {}
        unresolved: Counter = Counter()
        for s in scans:
            decade = self._decade(s.year)
            matched: Dict[int, Dict[str, Any]] = {}
            for raw in s.subjects:
                record = by_name.get(normalize_location_name(raw))
                if record is None:
                    unresolved[raw] += 1
                    continue
                if record["type"] in actor_types:
                    matched[record["o_id"]] = record
            for o_id, record in matched.items():
                st = stats.get(o_id)
                if st is None:
                    st = stats[o_id] = {
                        "o_id": o_id,
                        "name": record["name"],
                        "type": record["type"],
                        "items": 0,
                        "tagged": 0,
                        "by_decade": Counter(),
                        "by_country": Counter(),
                        "by_subset": Counter(),
                        "first_year": s.year,
                        "last_year": s.year,
                    }
                st["items"] += 1
                if s.is_tagged:
                    st["tagged"] += 1
                if decade:
                    st["by_decade"][decade] += 1
                for country in s.countries:
                    st["by_country"][country] += 1
                st["by_subset"][s.subset] += 1
                if s.year:
                    lo, hi = st["first_year"], st["last_year"]
                    st["first_year"] = s.year if lo is None else min(lo, s.year)
                    st["last_year"] = s.year if hi is None else max(hi, s.year)

        kept = [st for st in stats.values() if st["items"] >= self.min_actor_items]
        kept.sort(key=lambda st: (-st["items"], st["name"]))
        decades = sorted({d for st in kept for d in st["by_decade"]})

        actors = [{
            "o_id": st["o_id"],
            "name": st["name"],
            "type": st["type"],
            "items": st["items"],
            "tagged": st["tagged"],
            "first_year": st["first_year"],
            "last_year": st["last_year"],
            "by_decade": [st["by_decade"].get(d, 0) for d in decades],
            "by_country": dict(st["by_country"]),
            "by_subset": dict(st["by_subset"]),
        } for st in kept]

        self.logger.info(
            f"  actors: {len(actors)} of {len(stats)} authority records clear "
            f"{self.min_actor_items} items ({indexed} index records joined, "
            f"{len(unresolved)} subject strings unresolved)")
        return {
            "generated_at": generate_timestamp(),
            "min_items": self.min_actor_items,
            "types": sorted(actor_types),
            "decades": decades,
            "actors": actors,
            "index_records": indexed,
            "unresolved": [
                {"name": name, "count": n}
                for name, n in unresolved.most_common(20)
            ],
            "unresolved_total": len(unresolved),
            "note": (
                "Curated authority records co-occurring with the dossier's "
                "items, counted once per item. Subject headings are excluded: "
                "they are catalogued research vocabulary, not actors."
            ),
        }
